### ISSUE TITLE
Pipeline plugin compatibility (Formerly Workflow plugin)

### ISSUE THREAD

**Issue by leth** [jenkinsci/stashnotifier-plugin#92]
I'd be great if we could use this plugin from a `Jenkinsfile`!

The plugin repo has [a compatibility guide](https://github.com/jenkinsci/workflow-plugin/blob/master/COMPATIBILITY.md#plugin-developer-guide).

**Comment by scaytrase** [comment]
How does this handle credentials?

**Comment by leth** [comment]
It looks like they reference them by ID, e.g:
```
git url: 'git@git.local.domain:ps/project.git', credentialsId: '67c3072d-b9a7-44fa-a5aa-560ba9c1662f'
```

**Comment by jglick** [comment]
Probably suffices to implement `SimpleBuildStep` and replace `AbstractBuild` with `Run` everywhere, skipping calls to `TokenMacro` if not in fact passed an `AbstractBuild`.

**Comment by lindt** [comment]
is there any known possible workaround?

Something like?
```
sh 'curl https://stash..../foo'
```

**Comment by scaytrase** [comment]
Sure, yes, you could just use official stash\bitbucket server documentation

https://developer.atlassian.com/stash/docs/latest/how-tos/updating-build-status-for-commits.html

**Comment by headcrabmeat** [comment]
Hi,
Does anyone is working on this? If not please assign it to me

**Comment by headcrabmeat** [comment]
BTW @scaytrase, 
whats the target branch should be? master or relase1.0 I see some refactoring on the master but new classes are not currently in use.

**Comment by scaytrase** [comment]
Use default branch (**release/1.x** currently) as target

**Comment by headcrabmeat** [comment]
To all who are interested in. 

I believe that the usage of stash notifier plugin in pipeline should be changed. 
I think that behavior can be same as in the email plugin. 
You can notify stash any time with any status for any repository you want.

So you add a step in the beginning of the build section and insert

``` groovy
step([$class: 'StashNotifier', buildState: 'INPROGRESS', commitSha1: 'bbbb', credentialsId: '', disableInprogressNotification: false, ignoreUnverifiedSSLPeer: false, includeBuildNumberInKey: false, prependParentProjectKey: false, projectKey: 'cccc', stashServerBaseUrl: 'aaaa'])

```

On failure you just catch the error
like

``` groovy
stage 'build'
try {
     sh 'do the build'
} catch (e) {
   step([$class: 'StashNotifier', buildState: 'FAILED', commitSha1: 'bbbb', credentialsId: '', disableInprogressNotification: false, ignoreUnverifiedSSLPeer: false, includeBuildNumberInKey: false, prependParentProjectKey: false, projectKey: 'cccc', stashServerBaseUrl: 'aaaa'])
error e
}
```

On success you send
``` groovy
step([$class: 'StashNotifier', buildState: 'SUCCESS', commitSha1: 'bbbb', credentialsId: '', disableInprogressNotification: false, ignoreUnverifiedSSLPeer: false, includeBuildNumberInKey: false, prependParentProjectKey: false, projectKey: 'cccc', stashServerBaseUrl: 'aaaa'])
````

Overall the snippet generator might look like this
![image](https://cloud.githubusercontent.com/assets/1569458/14911952/96e29b94-0e01-11e6-9b87-e7b6f0ffd5e1.png)

I'm free for any other ideas. Just let me know.

**Comment by scaytrase** [comment]
Could you post a sample build file to represent the case-branching model (inprogress, fail, success)

**Comment by headcrabmeat** [comment]
@scaytrase, sure, will do it later. Now I have to get back to work.

**Comment by scaytrase** [comment]
Okey, just wan't to see how it will look like to give opinion. I've personally dislike the idea to store so much information in scripts (too much coupling to the infrastructue)

**Comment by nickbroon** [comment]
There appears to be Github Commit Status Notifier step that seems very similar, that can can be used from the Snippet editor.

step([$class: 'GitHubCommitStatusSetter', commitShaSource: [$class: 'ManuallyEnteredShaSource', sha: '123456'], statusResultSource: [$class: 'ConditionalStatusResultSource', results: [[$class: 'AnyBuildResult', message: 'Foo Bar', state: 'FAILURE']]]])

https://github.com/jenkinsci/github-plugin/blob/master/src/main/java/com/cloudbees/jenkins/GitHubSetCommitStatusBuilder.java

It would probably be good to model it on that.

**Comment by headcrabmeat** [comment]
@scaytrase 
I keep in mind following snippet:

```groovy
node('master') {
    _WORKSPACE=pwd()
    for ( repository in ['repository1', 'repository2', 'repository3'] ) {
        
        stage repository+'-verification'    // repository1-verification
        
        dir(_WORKSPACE+'/'+repository) {
             
             git 'https://mybb.org/project/reposytory1.git'

             // Somehow get commit sha1             
             notify('repository','aabbcc','INPROGRESS')

             try {
                 echo 'do the build/test/whatever'
             } catch (e) {
                   notify('repository','aabbcc','FAILED')
               error e
             }
             notify('repository','aabbcc','SUCCESS')
         }
    }
}

def notify(repository,sha1,status) {
    step([$class: 'StashNotifier', 
    buildState: status, 
    commitSha1: sha1, 
    credentialsId: '1111-2222-3333-4444', 
    disableInprogressNotification: false, 
    ignoreUnverifiedSSLPeer: false, 
    includeBuildNumberInKey: false, 
    prependParentProjectKey: false, 
    projectKey: 'myproject',
    repositoryKey: repository,     // new argument as it's not clear how to get git url in workflow
    stashServerBaseUrl: 'https://mybb.org'])
 }

```

**Comment by scaytrase** [comment]
> repositoryKey: repository,     // new argument as it's not clear how to get git url in workflow

What for? **StashBaseUrl** + **CommitSha1** should be enought to notify.

**Comment by nowayride** [comment]
How do we even derive the commit hash from pipelines? JENKINS-34455 gives a workaround using shell but pipelines don't expose a commit hash variable from what I can see, even though the job recognizes one.

**Comment by scaytrase** [comment]
When building classic build steps, commit sha1 is known from the ENV. When using suggested 

```groovy
 git "https://mybb.org/project/${repository}.git"
```

it could be obtained from workdir, i think

**Comment by tmds** [comment]
@headcrabmeat, any idea when we can expect this feature?

**Comment by tmds** [comment]
Is someone working on this issue? Is there a workaround?

**Comment by scaytrase** [comment]
Workaround is proposed to pipeline plugin

https://github.com/jenkinsci/pipeline-examples/pull/21

**Comment by headcrabmeat** [comment]
@tmds
It's hard to say. I still do not have a working code. All I can say is that I'm motivated to implement it's asap, as I want to use a plugin instead of workaround.

**Comment by tmds** [comment]
@headcrabmeat good to hear you are still on the issue! Depending on where you are it may be usefull to initiate a pull request and get some feedback even if things aren't working yet.

**Comment by headcrabmeat** [comment]
@tmds, @scaytrase Creatted PR #108 
Comments are welcome

**Comment by tmds** [comment]
@scaytrase can you review the PR?

**Comment by scaytrase** [comment]
I think there is nothing to review until tests pass. Everything could change. I'm keeping eye on it, don't worry. The patch is full of temporary changes, I think @headcrabmeat let us know when he is ready

**Comment by tmds** [comment]
@headcrabmeat the pipeline-examples has an example to get the commit sha:
https://github.com/jenkinsci/pipeline-examples/blob/912c73791fdc1a6faa1ab159c4719e33eba18688/pipeline-examples/gitcommit/gitcommit.groovy

**Comment by scaytrase** [comment]
Not the best example, as for me. Very platform dependent, requires `sh`, `git` in path (Jenkins has `jgit` as option), creates temporary files

**Comment by tmds** [comment]
> Not the best example, as for me. Very platform dependent, requires sh, git in path (Jenkins has jgit as option), creates temporary files

Only example so far and good as a starting point.

**Comment by headcrabmeat** [comment]
Thanks guys, I was told that `env` object can be used as well. I will try it soon.

**Comment by coder-hugo** [comment]
I implemented such a build step a few weeks ago for the [gitlab-plugin](https://github.com/jenkinsci/gitlab-plugin). You can have a look at the [GitLabCommitStatusStep](https://github.com/jenkinsci/gitlab-plugin/blob/master/src/main/java/com/dabsquared/gitlabjenkins/workflow/GitLabCommitStatusStep.java) as an example how to do this.

**Comment by scaytrase** [comment]
Implemented with #115
